pkg/kubectl/cmd/attach.go

```diff
@@ -167,7 +167,7 @@ func (p *AttachOptions) Run() error {
 		if err != nil {
 			return err
 		}
-		if pod.Status.Phase != api.PodRunning {
+		if pod.Status.Phase != api.PodRunning && (pod.Status.Phase == api.PodPending && len(pod.Spec.InitContainers) < 1) {
```

Comment:
I mean, it seems that the expression doesn't cover all the cases. 
cc @smarterclayton